components/filter/filter.js

```diff
@@ -0,0 +1,259 @@
+				</d2l-dropdown-menu>
+			</d2l-dropdown-button-subtle>
+		`;
+	}
+
+	_buildDimension(dimension) {
+		let dimensionHTML;
+		switch (dimension.tagName.toLowerCase()) {
+			case 'd2l-filter-dimension':
+				dimensionHTML = this._createListDimension(dimension);
+				break;
+		}
+		return html`
+			<d2l-hierarchical-view
+				@d2l-hierarchical-view-show-complete="${this._handleDimensionShowComplete}"
+				@d2l-hierarchical-view-show-start="${this._handleDimensionShowStart}"
+				id="${dimension.key}">
+
+				${dimensionHTML}
+			</d2l-hierarchical-view>
+		`;
+	}
+
+	_buildDimensions() {
+		return Array.from(this._dimensions.values(), (dimension) => {
+			const builtDimension = this._buildDimension(dimension);
+			return html`<d2l-menu-item text="${dimension.text}">
+				${builtDimension}
+			</d2l-menu-item>`;
+		});
+	}
+
+	_createListDimension(dimension) {
+		const slot = dimension.shadowRoot.querySelector('slot');
+		const items = slot.assignedNodes().filter(node => node.nodeType ===  Node.ELEMENT_NODE);
+		return html`
+			<d2l-list
+				@d2l-list-selection-change="${this._handleChangeListDimension}"
+				extend-separators>
+
+				${items.map(item => html`
+					<d2l-list-item
+						key="${item.key}"
+						selectable
+						?selected="${item.selected}"
+						slim>
+
+						<div class="d2l-filter-dimension-value-text">${item.text}</div>
+					</d2l-list-item>
+				`)}
+			</d2l-list>
+		`;
+	}
+
+	_dispatchChangeEvent(eventDetail) {
+		this.dispatchEvent(new CustomEvent('d2l-filter-change', { bubbles: false, composed: false, detail: eventDetail }));
+	}
+
+	_handleChangeListDimension(e) {
+		const dimensionKey = e.composedPath()[0].parentNode.id;
+		const valueKey = e.detail.key;
+		const selected = e.detail.selected;
+
+		// Update the corresponding d2l-filter-dimension-value to keep them in sync
+		const dimension = this._dimensions.get(dimensionKey);
+		const slot = dimension.shadowRoot.querySelector('slot');
+		const items = slot.assignedNodes().filter(node => node.nodeType ===  Node.ELEMENT_NODE);
+		const item = items.find(item => item.key === valueKey);
+		item.selected = selected;
+
+		this._dispatchChangeEvent({ dimension: dimensionKey, value: { key: valueKey, selected: selected } });
+	}
+
+	_handleDataChangeListDimension(e) {
+		const dimension = this.shadowRoot.getElementById(e.target.key);
+		const changes = e.detail.changes;
+		if (changes.has('text')) {
+			const menuItem = dimension.parentNode;
+			if (menuItem.text !== changes.get('text')) {
+				this.requestUpdate();
+				return;
+			}
+		}
+	}
+
+	_handleDataChangeListDimensionValue(e) {
+		const dimension = this.shadowRoot.getElementById(e.target.parentNode.key);
+		const value = dimension.querySelector(`[key="${e.target.key}"]`);
+		const changes = e.detail.changes;
+
+		if (changes.has('text')) {
+			const text = value.querySelector('.d2l-filter-dimension-value-text');
+			if (text.innerText !== changes.get('text')) {
+				text.innerText = changes.get('text');
```

Comment:
This ends up modifying the DOM outside of `render()`, which we try to avoid. Could we not just `requestUpdate` here which will cause a re-render and the `buildDimension` methods will re-create everything? Lit should then be smart enough to know if there were actually any changes and only make those that were necessary.

Reply:
Yeah that's what this wordy comment was trying to ask: https://github.com/BrightspaceUI/core/pull/1301#discussion_r622330195
Sounds like you're leaning towards 1 - avoid modifying the DOM outside of render and let Lit handle things in a smart well, but still check if we can avoid it altogether?

Reply:
I think so yeah. As you get the events, you should still be able to compare the new value against what's already in state and then only call `requestUpdate` if they're different. Checking an internal JSON object should be a lot cleaner/quicker than searching through DOM!